keps/sig-scheduling/3902-decoupled-taint-manager/README.md

```diff
@@ -0,0 +1,272 @@
+
+## Summary
+
+In Kubernetes, `NodeLifecycleController` applies predefined `NoExecute` taints (e.g., `Unreachable`, `NotReady`) when the nodes are determined to be unhealthy. After the nodes get tainted, `TaintManager` does its due diligence to start deleting running pods on those nodes based on `NoExecute` taints, which can be added by anyone.
+
+In this KEP, we propose to decouple `TaintManager` that performs taint-based pod eviction from `NodeLifecycleController` and make them two separate controllers: `NodeLifecycleController` to add taints to unhealthy nodes and `TaintManager` to perform pod deletion on nodes tainted with NoExecute effect.
+
+This separation not only improves code organization but also makes it easier to improve `TaintManager` or build custom `TaintManager`.
+
+## Motivation
+`NodeLifecycleController` combines two independent functions: 
+   * Adding a pre-defined set of `NoExecute` taints to nodes based on the node conditions 
+   * Performing pod eviction on `NoExecute` taints
+
+Splitting the `NodeLifecycleController` based on the above functionalities will help to disentangle code and make future extensions to either component manageable.
+
+### Goals
+
+* Move taint-based eviction implementation out of `NodeLifecycleController` into a separate and independent taint-manager to enhance separation of concerns, maintainability. 
+
+### Non-Goals
+
+The main focus of the KEP is on the separation of concerns between `NodeLifecycleController` and `TaintManager`, which will improve code organization and enable future extensions to `TaintManager`. While it is true that the separation allows cluster operators to build custom `TaintManager` and disable the default `TaintManager`, this is a side-effect of the change rather than the main goal. the KEP emphasizes the benefits of the separation of concerns and briefly mention the option to disable the default TaintManager as a potential side-effect.
+
+The non-goals are the following.
+* Introduce extra maintenance burden or regression after the code reorganization
+* Introduce incompatible behavior of `NodeLifecycleController` or `TaintManager`
+* Extend and enhance the current built-in APIs for customization of node taint-based eviction, including additional fields, more flexible communication protocols, webhooks and CRD references, etc.
+* Actual extension and enhancement  of the current `TaintManager`.
+
+## Proposal
+
+### User Stories (Optional)
+
+While this split is for improving the code maintainability, an effect of this change is that it allows cluster-admins to extend and enhance the default `TaintManager` and even replace the default `TaintManager` with a custom implementation. However, discussing such use-cases for extending `TaintManager` or writing  custom `TaintManager` is beyond the scope of this KEP. 
+
+#### Story
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
```

Comment:
I guess the risk here is that it will be harder to implement scenarios that need both features to be in the same controller. Just dreaming out loud those might be:

1. Fast notification between these two controllers to minimize time between taint being applied and eviction started. 
2. Cancellation of Pods eviction.
3. Pod eviction logic un-tainting the Node based on some signal received during the evicton.

I think mitigation for all of them would be the fact that we likely don't want this behavior built-in. Worth mentioning I think

Reply:
Good point! How about adding the following

"
Compared with a single combined controller, the risks of using two separate controllers include the following.
* Slightly increase the communication overhead from applying node taints to performing pod eviction.
* Make cancellation of pod eviction and un-tainting nodes harder.

It's a tradeoff between using the built-in simple controller and customizing the behavior to meet advanced needs.
"

Reply:
> It's a tradeoff between using the built-in simple controller and customizing the behavior to meet advanced needs

I do not agree with the statement that built-in is necessarily simple and separation is the way to meet the advanced needs. Combined controller can be configureable. I think justification is more along the lines of cleaner design is better as we do not believe those scenarios are very important

Reply:
> > It's a tradeoff between using the built-in simple controller and customizing the behavior to meet advanced needs
> 
> I do not agree with the statement that built-in is necessarily simple and separation is the way to meet the advanced needs. Combined controller can be configureable. I think justification is more along the lines of cleaner design is better as we do not believe those scenarios are very important

How about the following? 

`
While the performance impact of the split controllers is expected to be minimal, and pod eviction cancellation and un-tainting nodes are not common use cases, splitting  `NodeLifecycleController` promotes a cleaner design and enhances the overall maintainability and extensibility of in the long term.
`

Reply:
Thanks, Sergey. May I assume the KEP has been approved by you (sig-node)?